---
layout: default
title: {{ site.name }}
---

<p><b>EMASE-Zero</b> is a C++ implementation of EMASE, Expectation-Maximization (EM) algorithm for Allele Specific Expression (ASE). Unlike other approaches, we take into account the nested relationship of isoforms and alleles under genes when we disambiguate multiply aligning reads, which improves downstream analyses performed at the gene level. In combination with <a href="churchill-lab.github.io/alntools">alntools</a> (an alignment file processor that generates compressed three-dimensional incidence matrix), Zero estimates the expected read counts fast, over 10 times faster than RSEM. Zero also generalizes our fast hierarchical EM to any decent alignment strategies. The original python implementation is available at <a href="https://github.com/churchill-lab/emase">https://github.com/churchill-lab/emase</a>.</p>

<ul><li>Free software: MIT license</li></ul>

<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Building and installing EMASE-Zero requires a C++ compiler. After unzipping or extracting the downloaded file, please do the following:</p>

<pre>
  <code>$ cd src
  $ make</code>
</pre>

<p>This will create an executable called emase-zero in the src directory. You can copy this executable into your desired bin directory (such as /usr/local/bin).</p>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>EMASE-Zero takes in an alignment incidence matrix file (a binary format) generated by <a href="churchill-lab.github.io/alntools">alntools</a> using the diploid transcriptome alignment BAM file and two text files containing transcript lengths and gene-to-transcript mapping information. We need to specify which <a href="models">EMASE Model</a> to use for disambiguating the origins of multiply-aligning reads.</p>

<pre>
  <code>$ emase-zero -h

  EMASE-Zero Version x.y.z Help
  -----------------------------

  USAGE: emase-zero [options] <alignment_incidence_file>

  OPTIONS
    --bin (-b):
        Alignment incident matrix file (Binary format) prepared with alntools
        (https://churchill-lab.github.io/alntools)

    --outbase (-o) <filename>:
        Specify basename for output file (default="emase-zero.quantified")

    --model (-m) <int>:
        Specify EMASE model (can be 1-4, default=4)

    --transcript-lengths (-l) <filename>:
        Filename for transcript length file. Tab delimited file where the first
        field is the transcript ID, and the second is transcript length

    --gene-mappings (-g) <filename>:
        Filename containing transcript to gene mapping. Tab delimited file where
        the first field is the gene ID, all other fields are the transcript IDs
        that belong to that gene

    --max-iterations (-i) <int>:
        Specify the maximum number of EM iterations (Default=9999)

    --tolerance (-t) <double>:
        Specify the convergence threshold. EM will terminate when the
        sum of the aboslute value of differences in the total TPM from one
        iteration to the next is lower than this value if performing length
        adjustment and multipled by #reads if not (Default=0.000001)

    --verbose (-v):
        Run in verbose mode

    --version:
        Print the version and exit

    --help (-h):
        Print this message and exit</code>
</pre>

<p>&#8226;<code><strong>-b</strong></code> or <code><strong>--bin</strong></code>: Specify the name of binary equivalence class (EC) file. A binary EC file is generated from a bam file using <a href="churchill-lab.github.io/alntools">alntools</a>. EMASE-Zero does not require any specific aligner to use, but we recommend <a href="http://bowtie-bio.sourceforge.net/index.shtml">bowtie</a> Please see example below for the bowtie options we use for RNA-Seq analyses.</p>

<p>&#8226;<code><strong>-m</strong></code> or <code><strong>--model</strong></code>: Specify EMASE model for disambiguating multireads. According to our benchmarking, Model 2 worked best for ASE while Model 4 were the best for total gene exression.
<p><a id="models" href="#models"></a><img src="EMASE_models.jpg" alt="" title="EMASE models"></p>

<p>&#8226;<code><strong>-t</strong></code> or <code><strong>--transcript-lengths</strong></code>: Specify transcript length file name. We assume the file has the format of "Transcript_ID[tab]Length". For example, assuming 'M' and 'P' stand for maternal and paternal alleles respectively,</p>
<pre><code>ENSMUST00000000001_M	3262
...
ENSMUST00000098609_M	2653
...
ENSMUST00000000001_P	3264
...
ENSMUST00000098609_P	2659
...</code>
</pre>

<p>&#8226;<code><strong>-g</strong></code> or <code><strong>--gene-mappings</strong></code>: Specify gene-to-transcripts mapping file name. The file should be formatted in Gene_ID[tab]Transcript_ID_1[tab]Transcript_ID_2[tab]&hellip;. These mappings are among reference IDs, not among allele IDs, so no haplotype code at the end like _M or _P. For example,</p>
<pre><code>ENSMUSG00000000001	ENSMUST00000000001
ENSMUSG00000000003	ENSMUST00000000003	ENSMUST00000114041
ENSMUSG00000000028	ENSMUST00000000028	ENSMUST00000096990	ENSMUST00000115585
...</code>
</pre>

<p>Here's an example pipeline of commands:</p>
<pre><code>$ bowtie -q -a --best --strata --sam -v 3 -m 100 bowtie.transcriptome my_sample.fastq | samtools view -bS - > my_sample.bam
$ alntools bam2ec -t transcripts.lengths.tsv -c 8 my_sample.bam my_sample.bin
$ emase-zero -m 2 -b my_sample.bin -o emase-zero.quantified -l transcript.lengths.tsv -g gene2transcripts.tsv</code>
</pre>

<p> Note: We highly recommend adding poly-A tail to the alignment target sequences (transcripts) when you build bowtie index. We usually add (R-1)-long poly-A tails where R is the read length.</p>

<br>
<div id="home">
  <h2>Announcements</h2>
  <ul class="posts">
    {% for post in site.posts %}
      <li><span>{{ post.date | date_to_string }}</span> &raquo; <a href="{{ site.github.url }}{{ post.url }}">{{ post.title }}</a></li>
    {% endfor %}
  </ul>
</div>
